### PR TITLE
Update deps, add Scala 3 support, drop Scala 2.12

### PR DESCRIPTION
**Pull request by migesok**
Compared to safe-akka 3.0.0 the changes are source and binary compatible with the exception of these dependency changes:
- com.evolutiongaming:executor-tools_2.13: missing dependency
- com.google.protobuf:protobuf-java: missing dependency
- org.scala-lang.modules:scala-java8-compat_2.13: incompatible version change from 0.9.0 to 1.0.0 (compatibility: early semantic versioning)
- org.typelevel:cats-macros_2.13: missing dependency

Changes highlights:
- updated plugins and dependencies, dropped unused dependencies
- added Scala 3.3.0 cross-compilation
- removed Scala 2.12 support
- introduced sbt-version-policy plugin and MiMa bincompat checks
- com.github.dnvriend:akka-persistence-inmemory is abandoned and doesn't have a Scala 3 version. Since I couldn't find a maintained replacement library with both Akka 2.6 and Scala 3 support, it was replaced with a combination of internal Akka in-mem journal plugin and a custom in-mem snapshot plugin implementation.
- fixed compilation warnings, minor test refactoring

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,22 +1,48 @@
-import Dependencies._
+import Dependencies.*
+
+// Your next release will be binary compatible with the previous one,
+// but it may not be source compatible (ie, it will be a minor release).
+ThisBuild / versionPolicyIntention := Compatibility.BinaryCompatible
+
+//TODO: after 3.1.0 release - clear up versionPolicyIgnored section
```
Comment (mr-git) [inline build.sbt:7]: let's ticket this TODO

File: build.sbt [hunk 2]
```diff
@@ -1,22 +1,48 @@
...
+  //com.evolutiongaming:executor-tools_2.13: missing dependency
+  "com.evolutiongaming" %% "executor-tools",
+  //com.google.protobuf:protobuf-java: missing dependency
+  "com.google.protobuf" % "protobuf-java",
+  //org.scala-lang.modules:scala-java8-compat_2.13:
+  //  incompatible version change from 0.9.0 to 1.0.0 (compatibility: early semantic versioning
+  "org.scala-lang.modules" %% "scala-java8-compat",
+  //org.typelevel:cats-macros_2.13: missing dependency
+  "org.typelevel" %% "cats-macros",
+)
```
Comment (mr-git) [inline build.sbt:22]: I suppose that with next version we will be able to remove these - let's add to above TODO ticket too

File: safe-actor/src/main/scala/com/evolutiongaming/safeakka/actor/SafeActorRef.scala [hunk 3]
```diff
@@ -53,7 +53,12 @@ object SafeActorRef {
     def unsafe = ref.unsafe
     def tell[C](msg: C, sender: Option[ActorRef])(implicit canRcv: CanRcv[C, A]) = canRcv match {
       case CanRcv.Identity        => ref.tell(f(msg.asInstanceOf[A]), sender)(CanRcv.identity)
-      case canRcv: CanRcv.Sys[C] => ref.tell(msg, sender)(canRcv)
+      //workaround for Scala 3.3.0 compiler quirk:
+      //  asInstanceOf added because of a compilation error on pattern matching,
+      //  it complained about non-exhaustive match but suggested cases were non-existing
+      //
+      //old code: case canRcv: CanRcv.Sys[C] => ref.tell(msg, sender)(canRcv)
+      case canRcv: CanRcv.Sys[?] => ref.tell(msg, sender)(canRcv.asInstanceOf[CanRcv.Sys[C]])
```
Comment (mr-git) [inline safe-actor/src/main/scala/com/evolutiongaming/safeakka/actor/SafeActorRef.scala:61]: in https://users.scala-lang.org/t/alternative-for-type-ascriptions-after-tuple-patterns-in-scala-3-3-0/9328 I found alternative:
```suggestion
      case canRcv: CanRcv.Sys[C@unchecked] => ref.tell(msg, sender)(canRcv)
```

seems that it works in other cases too

File: safe-actor/src/test/scala/com/evolutiongaming/safeakka/actor/SafeActorSpec.scala [hunk 4]
```diff
@@ -110,7 +110,11 @@ class SafeActorSpec extends AnyWordSpec with ActorSpec with Matchers {
     val setup: SetupActor[Cmd] = ctx => {
       val behavior = Behavior.stateless[Cmd] {
         case Signal.Msg(cmd, sender)   =>
-          cmd match {
+          //workaround for Scala 3.3.0 compiler quirk:
+          //  typedCmd assignment added because of a compilation error on pattern matching,
+          //  it complained about non-exhaustive match but suggested cases were non-existing
+          val typedCmd: Cmd = cmd
+          typedCmd match {
```
Comment (mr-git) [inline safe-actor/src/test/scala/com/evolutiongaming/safeakka/actor/SafeActorSpec.scala:117]: I tried locally - old code compiles 🤷

File: safe-persistence-async/src/main/scala/com/evolutiongaming/safeakka/persistence/async/AsyncPersistentBehavior.scala [hunk 5]
```diff
@@ -90,6 +89,7 @@ private[async] class AsyncPersistentBehavior[S, C, E](
       def onHandlers(hs: Nel[SignalAndHandler], promise: Promise[Unit]) = {
         type R = CmdResult.NonStop[S, E]
 
+        @nowarn("cat=deprecation")
```
Comment (mr-git) [inline safe-persistence-async/src/main/scala/com/evolutiongaming/safeakka/persistence/async/AsyncPersistentBehavior.scala:92]: why cannot we respect following deprecation?
```
  @deprecated("use `Change` instead", "3.0.0")
  final case class Records[+S, +E](
```

File: build.sbt [hunk 6]
```diff
@@ -1,22 +1,48 @@
...
+  organizationHomepage := Some(url("https://evolution.com")),
   scalaVersion := crossScalaVersions.value.head,
-  crossScalaVersions := Seq("2.13.4", "2.12.12"),
+  crossScalaVersions := Seq("2.13.14", "3.3.3"),
+  Compile / scalacOptions ++= {
+    if (scalaBinaryVersion.value == "2.13") {
+      Seq(
+        "-Xsource:3"
+      )
+    } else Seq.empty
```
Comment (mr-git) [inline build.sbt:37]: our plugin for `scalac` doesn't add any options for 3.3.3 😞 

maybe we can add some manually here (or upgrade our plugin with good and safe set for Scala 3.3.3), like:
```suggestion
    } else Seq("-deprecation")
```

File: safe-actor/src/test/scala/com/evolutiongaming/safeakka/actor/SafeActorSpec.scala (hunk 4 above)
Comment (mr-git) [reply safe-actor/src/test/scala/com/evolutiongaming/safeakka/actor/SafeActorSpec.scala:117]: my bad - this indeed fails - `compile` doesn't compile tests have to use `Test/compile`

Comment (mr-git) [reply safe-actor/src/test/scala/com/evolutiongaming/safeakka/actor/SafeActorSpec.scala:117]: there is an alternative:
```scala
(cmd: Cmd) match {
```

File: build.sbt (hunk 1 above)
Comment (migesok) [reply build.sbt:7]: I will clean it up right after release, not sure it needs additional tickets.

File: build.sbt (hunk 6 above)
Comment (FunFunFine) [reply build.sbt:37]: we should add sbt-tpolecat instead. Not sure if it's worth doing in scope of this PR though

File: build.sbt (hunk 2 above)
Comment (migesok) [reply build.sbt:22]: It's covered by the TODO above, I believe.

**Review by FunFunFine — APPROVED** [review]

File: build.sbt (hunk 6 above)
Comment (mr-git) [reply build.sbt:37]: probably, just have to review it and either port the changes to our one or switch all our OSS to it

File: safe-actor/src/main/scala/com/evolutiongaming/safeakka/actor/SafeActorRef.scala (hunk 3 above)
Comment (migesok) [reply safe-actor/src/main/scala/com/evolutiongaming/safeakka/actor/SafeActorRef.scala]: Applied where it fits

File: safe-persistence-async/src/main/scala/com/evolutiongaming/safeakka/persistence/async/AsyncPersistentBehavior.scala (hunk 5 above)
Comment (migesok) [reply safe-persistence-async/src/main/scala/com/evolutiongaming/safeakka/persistence/async/AsyncPersistentBehavior.scala:92]: `CmdResult.Records` case has to be handled in the implementation or the behaviour will be different.

File: build.sbt (hunk 6 above)
Comment (migesok) [reply build.sbt:37]: I'd rather leave it to the Evolution scalac opts plugin:
https://github.com/evolution-gaming/sbt-scalac-opts-plugin/issues/9

The projects will get the options with the plugin update.

**Pull request merged**